### DISCUSSION TITLE
How can I use the C backend?

### DISCUSSION THREAD

**Discussion by pizzadox9999** [konsoletyper/teavm#770] [General]
Hi,

which steps I have to undertake to generate C code from my Java source code?

pizzadox9999

**Comment by konsoletyper** [comment]
There's no documentation for C backend. You can look at examples, [here](https://github.com/konsoletyper/teavm/tree/master/samples/benchmark) and [here](https://github.com/konsoletyper/teavm/tree/master/samples/pi). If you read documentation, you can notice some mentions of C backend [here](https://teavm.org/docs/tooling/gradle.html). The best information about C backend is source code, don't afraid to read it.

**Reply by pizzadox9999** [reply]
Hi, thank you for the answer. After a short look it seems to me that only the gradle build supports the C backend. I tried it with maven and "<targetType>C</targetType>" but without the desired result of getting C code. Is it only possible with gradle, or do I have to read more source code(my arge enemy =)).
pizzadox9999

**Reply by konsoletyper** [reply]
C backend should be also supported in Maven

**Reply by pizzadox9999** [reply]
Sorry big maestro, it was all under "target\generated\js". My fault. Thank you!

**Comment by pizzadox9999** [comment]
Is it possible to disable unicode support? So that uchar.h and his friends are not needed? How big is the relience on the c standard library?

**Comment by konsoletyper** [comment]
No, there's no way do disable unicode support. I'm not sure what you call "big" and how to measure reliance. TeaVM, as well as any other JVM, relies no some platform-specific stuff, in case of TeaVM it's not just libc, but also some parts of POSIX and Windows API (like those to handle date/time and to manage memory). Actually, TeaVM was only tested with Windows and Linux. I think it can also work in MacOS and iOS (which are also POSIX) after a couple of small ifdefs, but since I don't have any Apple device, I can't patch relevant parts of runtime.

You can look at TeaVM runtime yourself and decide if it's big and whether you able to patch it for particular platform. Just look for `runtime.h` file and all files around. TeaVM simply copies these files into output directory. Anyway, it's rather strange that some platform does not even support classlib. What are trying to do? Run Java on embedded devices?

**Comment by pizzadox9999** [comment]
Kinda if the N64 counts as an embedded device. I just Straight thought that teavm would "just" translate Java to C, but that was a fulish assumption. But I will play around with it and look what I am able to accomplish.

**Reply by konsoletyper** [reply]
I guess TeaVM mostly uses `char16_t`, so you can define this type somewhere and ifdef include for your platform.

**Reply by konsoletyper** [reply]
C "just" translates Java to C, and `uchar.h` is a part of C11 standard. Whenever platform does not contain this include file, it does not actually support C.
